hippo.go

```diff
@@ -126,80 +143,128 @@ func (c *Client) Do(ctx context.Context, req *http.Request) ([]byte, error) {
 	c.lock.RLock()
-	defer c.lock.RUnlock()
+	sender := c.Sender
+	c.lock.RUnlock()
+
+	// compact the batch, grouping the same values together
+	batch = compactTagBatchPart(*batch)
+	batch.Sender = sender
+
+	batchParts, err := c.split(batch)
+	if err != nil {
+		return nil, err
+	}
+
+	ret := &SendBatchResult{
+		PartsTotal:   len(batchParts),
+		UpsertsTotal: len(batch.Upserts),
+		DeletesTotal: len(batch.Deletes),
+		BatchGUID:    "", // not known until we send the first part
+	}
+
+	for i, batchPart := range batchParts {
+		batchPart.BatchGUID = ret.BatchGUID
+
+		requestBytes, err := json.Marshal(batchPart)
+		if err != nil {
+			return ret, fmt.Errorf("Error JSON marshalling request batch - [%s] - underlying error: %s", ret, err)
+		}
+
+		req, err := c.NewTagBatchPartRequest("POST", url, requestBytes)
+		if err != nil {
+			return ret, fmt.Errorf("Error building request to %s - [%s] - underlying error: %s", url, ret, err)
+		}
+		responseBytes, err := c.Do(ctx, req)
+		if err != nil {
+			return ret, fmt.Errorf("Error POSTing populators to %s - [%s] - underlying error: %s", url, ret, err)
+		}
+
+		if i == 0 {
+			// first response returns the batch GUID, which we need to include in subsequent batches
+			apiResponse := APIServerResponse{}
+			if err := json.Unmarshal(responseBytes, &apiResponse); err != nil {
+				return ret, fmt.Errorf("Error unmarshalling API batch response - [%s] - underlying error: %s", ret, err)
+			}
+			if apiResponse.GUID == "" {
+				return ret, fmt.Errorf("API response did not include a GUID for subsequent batches - [%s] - underlying error: %s", ret, apiResponse.Error)
+			}
+			ret.BatchGUID = apiResponse.GUID
+		}
 
-	// swap out our local copy with a compacted one that ensures all criteria are grouped by value
-	tmp := compactTagBatchPart(*rFull)
-	tmp.Sender = c.Sender
-	rFull = &tmp
+		// update response
+		ret.PartsSent++
+		ret.UpsertsSent += len(batchPart.Upserts)
+		ret.DeletesSent += len(batchPart.Deletes)
 
+		// slow down the HTTP batches a bit to avoid rate limiting
+		time.Sleep(time.Second)
+	}
+
+	return ret, nil
+}
+
+// split a batch into parts small enough for sending through Kentik API
+func (c *Client) split(rFull *TagBatchPart) ([]TagBatchPart, error) {
 	encode := func(r *TagBatchPart) ([]byte, error) {
 		if b, err := json.Marshal(r); err != nil {
-			return nil, err
+			return nil, fmt.Errorf("Error JSON-marshalling batch: %s", err)
 		} else {
 			return b, nil
 		}
 	}
 
-	// If the size is small enough, just return here. Assume default case
-	init, err := encode(rFull)
+	serializedBytes, err := encode(rFull)
 	if err != nil {
-		return nil, 0, err
-	} else if len(init) < MAX_HIPPO_SIZE {
-		return [][]byte{init}, len(rFull.Upserts), nil
+		return nil, err
+	}
+
+	outgoingRequestSize := c.OutgoingRequestSize
+	if outgoingRequestSize <= 0 {
+		outgoingRequestSize = DEFAULT_MAX_HIPPO_SIZE
+	}
+
+	if len(serializedBytes) < outgoingRequestSize {
```

Comment:
Should we change this from `<` to `<=` to match the maximum size? This way we avoid the 2 parts, where the second part would be empty...

Reply:
I'd be happy to, but it wouldn't be. If it decides it needs 2 parts, it'll split the upserts between the batches:
https://github.com/kentik/gohippo/blob/392dd6fa99eb7fbfca436c07f6f8dce57f01f90e/hippo.go#L236-L237

Reply:
Sure, I was thinking more along the general aspect of thinking that 'max' means the maximum possible, inclusive, rather than exclusive.